File: scala-package/src/main/scala/com/lal/Weights.scala
package com.lal

// these are some objects we expect to use as inputs and outputs of key functions
import org.apache.spark.sql.DataFrame
import org.apache.spark.ml.linalg.DenseVector

// this is for the validation split
import org.apache.spark.ml.PredictionModel
import org.apache.spark.ml.param.{Param, ParamMap}
import org.apache.spark.ml.tuning.{ParamGridBuilder, TrainValidationSplit, TrainValidationSplitModel}
import org.apache.spark.ml.evaulation.{Evaluator, BinaryClassificationEvaluator, RegressionEvaluator}

// this is for the chosen model structure
import org.apache.spark.ml.Estimator
import org.apache.spark.ml.classification.GBTClassifier
import org.apache.spark.ml.regression.GBTRegressor

/**
  * A trait to represent a ''feature weighting process''.
  *
  *
  * @author Edward Turner
  * @todo Add additional documentation about this trait
  * @version 1.0
  */

// this is a private abstract class not for use on the user end side
sealed trait Weights{

  val featureCol: String
  val predictionCol: String

  // this paramgrid is define upon inheritance
  val paramGrid: Array[ParamMap]

  // the evaluator to use
  val evaluator: Any

  // this is the base model we want to use, which is defined upon inheritance
  val model: Any

  /**
    * @return Returns the optimized `model` of this weighting process.
    */
  def optModel(df: DataFrame): TrainValidationSplitModel = {
    """
      | This function will optimize the Estimator chosen for this class
    """.stripMargin

    val cv = new TrainValidationSplit()
      .setEstimator(model)
      .setEvaluator(evaluator)
      .setEstimatorParamMaps(paramGrid)
      .setTrainRatio(0.8)
      .setParallelism(2)

    cv.fit(df)

  }

  /**
    * @return Returns feature importance based on the weighting process
    */
  def getFeatureWeights(df: DataFrame): DenseVector = {
    val opt_model = this.optModel(df)

    val fitted_model = opt_model.transform(df)

    return fitted_model.featureImportances
  }

}

/**
  * A trait to represent a ''feature weighting process'' using gradient boosting process.
  *
  *
  * @author Edward Turner
  * @todo Add additional documentation about this trait
  * @version 1.0
  */
sealed trait GBRegWeights {

  val model = new GBTRegressor()

  val evaluator = new RegressionEvaluator()

  val paramGrid = new ParamGridBuilder()
    .addGrid(model.maxIter, Array((0 until 15).map(x => 100*(x + 1))))
    .addGrid(model.maxDepth, Array((0 until 15)))
    .addGrid(model.stepSize, Array((-15 to 0).map(x => Math.pow(10.0, x.toDouble))))
    .addGrid(model.subsamplingRate, Array((-15 to 0).map(x => Math.pow(10.0, x.toDouble))))
    .build()


}

sealed trait GBClasWeights {

  val evaulator = new BinaryClassificationEvaluator()

  val model = new GBTClassifier()

}

/**
  * A class to represent a ''feature weighting process'' using the Gradient-Boosting Tree regression task.
  *
  * {{{
  * val weighter = GBRegressorWeights("prediction", "label")
  * weighter.featureCol
  * weighter.predictionCol
  * }}}
  * @author Edward Turner
  * @todo Add additional documentation about this class
  * @version 1.0
  */
final case class GBRegressorWeights() extends Weights with GBRegWeights {

}


/**
  * A class to represent a ''feature weighting process'' using the Gradient-Boosting Tree classification task.
  *
  * {{{
  * val weighter = GBClassifierWeights("prediction", "label")
  * weighter.featureCol
  * weighter.predictionCol
  * }}}
  * @author Edward Turner
  * @todo Add additional documentation about this class
  * @version 1.0
  */
final case class GBClassifierWeights() extends Weights with GBClasWeights {


}

File: lal/spark/weights.py
import numpy as np

from .utils.asserts import AssertArgumentSparkDataFrame

from pyspark.ml.regression import GBTRegressor
from pyspark.ml.classification import GBTClassifier

from pyspark.ml.evaluation import BinaryClassificationEvaluator, RegressionEvaluator, MulticlassClassificationEvaluator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder


class _LGBMWeightsBase:

    assertor = AssertArgumentSparkDataFrame()

    evaluator = None

    model = None

    def __init__(self, feature_col, label_col):
        self.feature_col = feature_col
        self.label_col = label_col

    def get_space_grid(self):
        model = self.model

        if model is None:
            raise NotImplementedError("The model parameter is not set.")

        return ParamGridBuilder() \
            .addGrid(model.stepSize, (10.0 ** np.arange(-10.0, 1.0, 2.0)).tolist()) \
            .addGrid(model.subsamplingRate, (np.arange(2.0, 11.0, 2.0) / 10.0).tolist()) \
            .addGrid(model.maxDepth, [1, 3, 5, 7, 9, 11]) \
            .addGrid(model.maxIter, [100, 500, 1000, 5000, 10000])\
            .build()

    @assertor.assert_arguments
    def get_feature_importances(self, sdf):
        """

        :param sdf:
        :return:
        """

        evaluator = self.evaluator

        if evaluator is None:
            raise NotImplementedError("The evaluator parameter is not set.")

        space_grid = self.get_space_grid()

        model = self.model

        crossval = CrossValidator(estimator=model,
                                  estimatorParamMaps=space_grid,
                                  evaluator=evaluator,
                                  numFolds=5)

        cvModel = crossval.fit(sdf)

        return cvModel.bestModel.featureImportances


class GBMWeightRegressor(_LGBMWeightsBase):
    """
    This object will derive the feature importance weights based on a continuous output.  It will optimize the
    Gradient Boosting Model based on a regression metric, and then return the featureImportances based on the most
    optimized result.
    """
    def __init__(self, **kwargs):
        _LGBMWeightsBase.__init__(self, **kwargs)

        model = GBTRegressor(featuresCol=self.feature_col, labelCol=self.label_col)
        evaluator = RegressionEvaluator(labelCol=self.label_col, metricName='mse')

        self.model = model
        self.evaluator = evaluator


class GBMWeightBinaryClassifier(_LGBMWeightsBase):
    """
    This object will derive the feature importance weights based on binary output.  It will optimize the
    Gradient Boosting Model based on a classification metric, and then return the featureImportances based on the most
    optimized result.
    """
    def __init__(self, **kwargs):
        _LGBMWeightsBase.__init__(self, **kwargs)

        model = GBTClassifier(featuresCol=self.feature_col, labelCol=self.label_col)
        evaluator = BinaryClassificationEvaluator(labelCol=self.label_col, metricName='areaUnderROC')

        self.model = model
        self.evaluator = evaluator


class GBMWeightMultiClassifier(_LGBMWeightsBase):
    """
    This object will derive the feature importance weights based on a multiclass output.  It will optimize the
    Gradient Boosting Model based on a classification metric, and then return the featureImportances based on the most
    optimized result.
    """
    def __init__(self, **kwargs):
        _LGBMWeightsBase.__init__(self, **kwargs)

        model = GBTClassifier(featuresCol=self.feature_col, labelCol=self.label_col)
        evaluator = MulticlassClassificationEvaluator(labelCol=self.label_col, metricName='f1')

        self.model = model
        self.evaluator = evaluator

